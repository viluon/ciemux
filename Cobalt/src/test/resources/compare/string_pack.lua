local endians = { ">", "<" }
local paddings = { "", "1", "2", "4", "8", "16" }

local function is_eq(a, b)
	return a == b or (a ~= a and b ~= b)
end

local function dump(tbl)
	local out = {}
	for i = 1, #tbl do
		out[i] = (type(tbl[i]) == "string" and "%q" or "%s"):format(tbl[i]):gsub("\\000", "\\0")
	end
	return table.concat(out, ", ")
end

local function pack(fmt, ...)
	local ok, packed = pcall(string.pack, fmt, ...)
	if not ok then return packed end

	local unpacked = { pcall(string.unpack, fmt, packed) }
	local ok = table.remove(unpacked, 1)

	local hex = ("%02x"):rep(#packed):format(packed:byte(1, -1))
	if not ok then return ("%s[%s]"):format(hex, unpacked[1]) end

	local len = table.remove(unpacked)
	if len ~= #packed + 1 then return ("%s[%d ~= %d]"):format(hex, len, #packed + 1) end

	local input = { ... }
	local eq = #input == #unpacked
	for i = 1, #input do if not is_eq(unpacked[i], input[i]) then eq = false break end end

	if not eq then return ("%s[%s]"):format(hex, dump(unpacked)) end

	return hex
end

local function pack_all(fmt, ...)
	io.write(("string.pack(%q, %s) ="):format(fmt, dump({ ... })))
	for _, endian in pairs(endians) do
		for _, padding in pairs(paddings) do
			local p = "!" .. padding .. endian
			io.write(("\t%s%s"):format(p, pack(p .. fmt, ...)))
		end
	end
	io.write("\n")
end

local bytes = { 0x7f, -0x80, 0x01 }
local ubytes = { 0xff, 0x01 }

local shorts = { 0x7fff, -0x8000, 0x01 }
local ushorts = { 0xffff, 0x0001 }

local ints = { 0x7fffffff, 0x01, -0x80000000 }
local uints = { 0x0000000001, 0xffffffffff, 0xdeadbeef }

local floats = { math.huge, -math.huge, math.abs(0/0), 1, 2, 3 }

local strings = { "a", "abc", "abcde" }

print("= Integers =")
for _, x in pairs(bytes) do pack_all("b", x) end
for _, x in pairs(ubytes) do pack_all("B", x) end

for _, x in pairs(shorts) do pack_all("h", x) end
for _, x in pairs(ushorts) do pack_all("H", x) end

for _, x in pairs(ints) do pack_all("l", x) end
for _, x in pairs(ints) do pack_all("j", x) end

for _, x in pairs(uints) do pack_all("L", x) end
for _, x in pairs(uints) do pack_all("J", x) end

print("\n= Variable width Integers =")
for i = 0, 4 do for _, x in pairs(bytes) do pack_all("i" .. math.floor(2^i), x) end end

print("\n= Floats =")
for _, x in pairs(floats) do pack_all("f", x) end
for _, x in pairs(floats) do pack_all("d", x) end
for _, x in pairs(floats) do pack_all("n", x) end

print("\n= Strings =")
for _, x in pairs(strings) do pack_all("c5", x) end
for _, x in pairs(strings) do pack_all("z", x) end
for _, x in pairs(strings) do pack_all("s4", x) end
for _, x in pairs(strings) do pack_all("s1", x) end

print("\n= Padding =")
pack_all("BXjB", 0xff, 0x7f)
pack_all("BxB", 0xff, 0x7f)
