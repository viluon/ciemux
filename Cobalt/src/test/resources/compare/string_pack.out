= Integers =
string.pack("b", 127) =	!>7f	!1>7f	!2>7f	!4>7f	!8>7f	!16>7f	!<7f	!1<7f	!2<7f	!4<7f	!8<7f	!16<7f
string.pack("b", -128) =	!>80	!1>80	!2>80	!4>80	!8>80	!16>80	!<80	!1<80	!2<80	!4<80	!8<80	!16<80
string.pack("b", 1) =	!>01	!1>01	!2>01	!4>01	!8>01	!16>01	!<01	!1<01	!2<01	!4<01	!8<01	!16<01
string.pack("B", 255) =	!>ff	!1>ff	!2>ff	!4>ff	!8>ff	!16>ff	!<ff	!1<ff	!2<ff	!4<ff	!8<ff	!16<ff
string.pack("B", 1) =	!>01	!1>01	!2>01	!4>01	!8>01	!16>01	!<01	!1<01	!2<01	!4<01	!8<01	!16<01
string.pack("h", 32767) =	!>7fff	!1>7fff	!2>7fff	!4>7fff	!8>7fff	!16>7fff	!<ff7f	!1<ff7f	!2<ff7f	!4<ff7f	!8<ff7f	!16<ff7f
string.pack("h", -32768) =	!>8000	!1>8000	!2>8000	!4>8000	!8>8000	!16>8000	!<0080	!1<0080	!2<0080	!4<0080	!8<0080	!16<0080
string.pack("h", 1) =	!>0001	!1>0001	!2>0001	!4>0001	!8>0001	!16>0001	!<0100	!1<0100	!2<0100	!4<0100	!8<0100	!16<0100
string.pack("H", 65535) =	!>ffff	!1>ffff	!2>ffff	!4>ffff	!8>ffff	!16>ffff	!<ffff	!1<ffff	!2<ffff	!4<ffff	!8<ffff	!16<ffff
string.pack("H", 1) =	!>0001	!1>0001	!2>0001	!4>0001	!8>0001	!16>0001	!<0100	!1<0100	!2<0100	!4<0100	!8<0100	!16<0100
string.pack("l", 2147483647) =	!>000000007fffffff	!1>000000007fffffff	!2>000000007fffffff	!4>000000007fffffff	!8>000000007fffffff	!16>000000007fffffff	!<ffffff7f00000000	!1<ffffff7f00000000	!2<ffffff7f00000000	!4<ffffff7f00000000	!8<ffffff7f00000000	!16<ffffff7f00000000
string.pack("l", 1) =	!>0000000000000001	!1>0000000000000001	!2>0000000000000001	!4>0000000000000001	!8>0000000000000001	!16>0000000000000001	!<0100000000000000	!1<0100000000000000	!2<0100000000000000	!4<0100000000000000	!8<0100000000000000	!16<0100000000000000
string.pack("l", -2147483648) =	!>ffffffff80000000	!1>ffffffff80000000	!2>ffffffff80000000	!4>ffffffff80000000	!8>ffffffff80000000	!16>ffffffff80000000	!<00000080ffffffff	!1<00000080ffffffff	!2<00000080ffffffff	!4<00000080ffffffff	!8<00000080ffffffff	!16<00000080ffffffff
string.pack("j", 2147483647) =	!>000000007fffffff	!1>000000007fffffff	!2>000000007fffffff	!4>000000007fffffff	!8>000000007fffffff	!16>000000007fffffff	!<ffffff7f00000000	!1<ffffff7f00000000	!2<ffffff7f00000000	!4<ffffff7f00000000	!8<ffffff7f00000000	!16<ffffff7f00000000
string.pack("j", 1) =	!>0000000000000001	!1>0000000000000001	!2>0000000000000001	!4>0000000000000001	!8>0000000000000001	!16>0000000000000001	!<0100000000000000	!1<0100000000000000	!2<0100000000000000	!4<0100000000000000	!8<0100000000000000	!16<0100000000000000
string.pack("j", -2147483648) =	!>ffffffff80000000	!1>ffffffff80000000	!2>ffffffff80000000	!4>ffffffff80000000	!8>ffffffff80000000	!16>ffffffff80000000	!<00000080ffffffff	!1<00000080ffffffff	!2<00000080ffffffff	!4<00000080ffffffff	!8<00000080ffffffff	!16<00000080ffffffff
string.pack("L", 1) =	!>0000000000000001	!1>0000000000000001	!2>0000000000000001	!4>0000000000000001	!8>0000000000000001	!16>0000000000000001	!<0100000000000000	!1<0100000000000000	!2<0100000000000000	!4<0100000000000000	!8<0100000000000000	!16<0100000000000000
string.pack("L", 1099511627775) =	!>000000ffffffffff	!1>000000ffffffffff	!2>000000ffffffffff	!4>000000ffffffffff	!8>000000ffffffffff	!16>000000ffffffffff	!<ffffffffff000000	!1<ffffffffff000000	!2<ffffffffff000000	!4<ffffffffff000000	!8<ffffffffff000000	!16<ffffffffff000000
string.pack("L", 3735928559) =	!>00000000deadbeef	!1>00000000deadbeef	!2>00000000deadbeef	!4>00000000deadbeef	!8>00000000deadbeef	!16>00000000deadbeef	!<efbeadde00000000	!1<efbeadde00000000	!2<efbeadde00000000	!4<efbeadde00000000	!8<efbeadde00000000	!16<efbeadde00000000
string.pack("J", 1) =	!>0000000000000001	!1>0000000000000001	!2>0000000000000001	!4>0000000000000001	!8>0000000000000001	!16>0000000000000001	!<0100000000000000	!1<0100000000000000	!2<0100000000000000	!4<0100000000000000	!8<0100000000000000	!16<0100000000000000
string.pack("J", 1099511627775) =	!>000000ffffffffff	!1>000000ffffffffff	!2>000000ffffffffff	!4>000000ffffffffff	!8>000000ffffffffff	!16>000000ffffffffff	!<ffffffffff000000	!1<ffffffffff000000	!2<ffffffffff000000	!4<ffffffffff000000	!8<ffffffffff000000	!16<ffffffffff000000
string.pack("J", 3735928559) =	!>00000000deadbeef	!1>00000000deadbeef	!2>00000000deadbeef	!4>00000000deadbeef	!8>00000000deadbeef	!16>00000000deadbeef	!<efbeadde00000000	!1<efbeadde00000000	!2<efbeadde00000000	!4<efbeadde00000000	!8<efbeadde00000000	!16<efbeadde00000000

= Variable width Integers =
string.pack("i1", 127) =	!>7f	!1>7f	!2>7f	!4>7f	!8>7f	!16>7f	!<7f	!1<7f	!2<7f	!4<7f	!8<7f	!16<7f
string.pack("i1", -128) =	!>80	!1>80	!2>80	!4>80	!8>80	!16>80	!<80	!1<80	!2<80	!4<80	!8<80	!16<80
string.pack("i1", 1) =	!>01	!1>01	!2>01	!4>01	!8>01	!16>01	!<01	!1<01	!2<01	!4<01	!8<01	!16<01
string.pack("i2", 127) =	!>007f	!1>007f	!2>007f	!4>007f	!8>007f	!16>007f	!<7f00	!1<7f00	!2<7f00	!4<7f00	!8<7f00	!16<7f00
string.pack("i2", -128) =	!>ff80	!1>ff80	!2>ff80	!4>ff80	!8>ff80	!16>ff80	!<80ff	!1<80ff	!2<80ff	!4<80ff	!8<80ff	!16<80ff
string.pack("i2", 1) =	!>0001	!1>0001	!2>0001	!4>0001	!8>0001	!16>0001	!<0100	!1<0100	!2<0100	!4<0100	!8<0100	!16<0100
string.pack("i4", 127) =	!>0000007f	!1>0000007f	!2>0000007f	!4>0000007f	!8>0000007f	!16>0000007f	!<7f000000	!1<7f000000	!2<7f000000	!4<7f000000	!8<7f000000	!16<7f000000
string.pack("i4", -128) =	!>ffffff80	!1>ffffff80	!2>ffffff80	!4>ffffff80	!8>ffffff80	!16>ffffff80	!<80ffffff	!1<80ffffff	!2<80ffffff	!4<80ffffff	!8<80ffffff	!16<80ffffff
string.pack("i4", 1) =	!>00000001	!1>00000001	!2>00000001	!4>00000001	!8>00000001	!16>00000001	!<01000000	!1<01000000	!2<01000000	!4<01000000	!8<01000000	!16<01000000
string.pack("i8", 127) =	!>000000000000007f	!1>000000000000007f	!2>000000000000007f	!4>000000000000007f	!8>000000000000007f	!16>000000000000007f	!<7f00000000000000	!1<7f00000000000000	!2<7f00000000000000	!4<7f00000000000000	!8<7f00000000000000	!16<7f00000000000000
string.pack("i8", -128) =	!>ffffffffffffff80	!1>ffffffffffffff80	!2>ffffffffffffff80	!4>ffffffffffffff80	!8>ffffffffffffff80	!16>ffffffffffffff80	!<80ffffffffffffff	!1<80ffffffffffffff	!2<80ffffffffffffff	!4<80ffffffffffffff	!8<80ffffffffffffff	!16<80ffffffffffffff
string.pack("i8", 1) =	!>0000000000000001	!1>0000000000000001	!2>0000000000000001	!4>0000000000000001	!8>0000000000000001	!16>0000000000000001	!<0100000000000000	!1<0100000000000000	!2<0100000000000000	!4<0100000000000000	!8<0100000000000000	!16<0100000000000000
string.pack("i16", 127) =	!>0000000000000000000000000000007f	!1>0000000000000000000000000000007f	!2>0000000000000000000000000000007f	!4>0000000000000000000000000000007f	!8>0000000000000000000000000000007f	!16>0000000000000000000000000000007f	!<7f000000000000000000000000000000	!1<7f000000000000000000000000000000	!2<7f000000000000000000000000000000	!4<7f000000000000000000000000000000	!8<7f000000000000000000000000000000	!16<7f000000000000000000000000000000
string.pack("i16", -128) =	!>ffffffffffffffffffffffffffffff80	!1>ffffffffffffffffffffffffffffff80	!2>ffffffffffffffffffffffffffffff80	!4>ffffffffffffffffffffffffffffff80	!8>ffffffffffffffffffffffffffffff80	!16>ffffffffffffffffffffffffffffff80	!<80ffffffffffffffffffffffffffffff	!1<80ffffffffffffffffffffffffffffff	!2<80ffffffffffffffffffffffffffffff	!4<80ffffffffffffffffffffffffffffff	!8<80ffffffffffffffffffffffffffffff	!16<80ffffffffffffffffffffffffffffff
string.pack("i16", 1) =	!>00000000000000000000000000000001	!1>00000000000000000000000000000001	!2>00000000000000000000000000000001	!4>00000000000000000000000000000001	!8>00000000000000000000000000000001	!16>00000000000000000000000000000001	!<01000000000000000000000000000000	!1<01000000000000000000000000000000	!2<01000000000000000000000000000000	!4<01000000000000000000000000000000	!8<01000000000000000000000000000000	!16<01000000000000000000000000000000

= Floats =
string.pack("f", inf) =	!>7f800000	!1>7f800000	!2>7f800000	!4>7f800000	!8>7f800000	!16>7f800000	!<0000807f	!1<0000807f	!2<0000807f	!4<0000807f	!8<0000807f	!16<0000807f
string.pack("f", -inf) =	!>ff800000	!1>ff800000	!2>ff800000	!4>ff800000	!8>ff800000	!16>ff800000	!<000080ff	!1<000080ff	!2<000080ff	!4<000080ff	!8<000080ff	!16<000080ff
string.pack("f", nan) =	!>7fc00000	!1>7fc00000	!2>7fc00000	!4>7fc00000	!8>7fc00000	!16>7fc00000	!<0000c07f	!1<0000c07f	!2<0000c07f	!4<0000c07f	!8<0000c07f	!16<0000c07f
string.pack("f", 1) =	!>3f800000	!1>3f800000	!2>3f800000	!4>3f800000	!8>3f800000	!16>3f800000	!<0000803f	!1<0000803f	!2<0000803f	!4<0000803f	!8<0000803f	!16<0000803f
string.pack("f", 2) =	!>40000000	!1>40000000	!2>40000000	!4>40000000	!8>40000000	!16>40000000	!<00000040	!1<00000040	!2<00000040	!4<00000040	!8<00000040	!16<00000040
string.pack("f", 3) =	!>40400000	!1>40400000	!2>40400000	!4>40400000	!8>40400000	!16>40400000	!<00004040	!1<00004040	!2<00004040	!4<00004040	!8<00004040	!16<00004040
string.pack("d", inf) =	!>7ff0000000000000	!1>7ff0000000000000	!2>7ff0000000000000	!4>7ff0000000000000	!8>7ff0000000000000	!16>7ff0000000000000	!<000000000000f07f	!1<000000000000f07f	!2<000000000000f07f	!4<000000000000f07f	!8<000000000000f07f	!16<000000000000f07f
string.pack("d", -inf) =	!>fff0000000000000	!1>fff0000000000000	!2>fff0000000000000	!4>fff0000000000000	!8>fff0000000000000	!16>fff0000000000000	!<000000000000f0ff	!1<000000000000f0ff	!2<000000000000f0ff	!4<000000000000f0ff	!8<000000000000f0ff	!16<000000000000f0ff
string.pack("d", nan) =	!>7ff8000000000000	!1>7ff8000000000000	!2>7ff8000000000000	!4>7ff8000000000000	!8>7ff8000000000000	!16>7ff8000000000000	!<000000000000f87f	!1<000000000000f87f	!2<000000000000f87f	!4<000000000000f87f	!8<000000000000f87f	!16<000000000000f87f
string.pack("d", 1) =	!>3ff0000000000000	!1>3ff0000000000000	!2>3ff0000000000000	!4>3ff0000000000000	!8>3ff0000000000000	!16>3ff0000000000000	!<000000000000f03f	!1<000000000000f03f	!2<000000000000f03f	!4<000000000000f03f	!8<000000000000f03f	!16<000000000000f03f
string.pack("d", 2) =	!>4000000000000000	!1>4000000000000000	!2>4000000000000000	!4>4000000000000000	!8>4000000000000000	!16>4000000000000000	!<0000000000000040	!1<0000000000000040	!2<0000000000000040	!4<0000000000000040	!8<0000000000000040	!16<0000000000000040
string.pack("d", 3) =	!>4008000000000000	!1>4008000000000000	!2>4008000000000000	!4>4008000000000000	!8>4008000000000000	!16>4008000000000000	!<0000000000000840	!1<0000000000000840	!2<0000000000000840	!4<0000000000000840	!8<0000000000000840	!16<0000000000000840
string.pack("n", inf) =	!>7ff0000000000000	!1>7ff0000000000000	!2>7ff0000000000000	!4>7ff0000000000000	!8>7ff0000000000000	!16>7ff0000000000000	!<000000000000f07f	!1<000000000000f07f	!2<000000000000f07f	!4<000000000000f07f	!8<000000000000f07f	!16<000000000000f07f
string.pack("n", -inf) =	!>fff0000000000000	!1>fff0000000000000	!2>fff0000000000000	!4>fff0000000000000	!8>fff0000000000000	!16>fff0000000000000	!<000000000000f0ff	!1<000000000000f0ff	!2<000000000000f0ff	!4<000000000000f0ff	!8<000000000000f0ff	!16<000000000000f0ff
string.pack("n", nan) =	!>7ff8000000000000	!1>7ff8000000000000	!2>7ff8000000000000	!4>7ff8000000000000	!8>7ff8000000000000	!16>7ff8000000000000	!<000000000000f87f	!1<000000000000f87f	!2<000000000000f87f	!4<000000000000f87f	!8<000000000000f87f	!16<000000000000f87f
string.pack("n", 1) =	!>3ff0000000000000	!1>3ff0000000000000	!2>3ff0000000000000	!4>3ff0000000000000	!8>3ff0000000000000	!16>3ff0000000000000	!<000000000000f03f	!1<000000000000f03f	!2<000000000000f03f	!4<000000000000f03f	!8<000000000000f03f	!16<000000000000f03f
string.pack("n", 2) =	!>4000000000000000	!1>4000000000000000	!2>4000000000000000	!4>4000000000000000	!8>4000000000000000	!16>4000000000000000	!<0000000000000040	!1<0000000000000040	!2<0000000000000040	!4<0000000000000040	!8<0000000000000040	!16<0000000000000040
string.pack("n", 3) =	!>4008000000000000	!1>4008000000000000	!2>4008000000000000	!4>4008000000000000	!8>4008000000000000	!16>4008000000000000	!<0000000000000840	!1<0000000000000840	!2<0000000000000840	!4<0000000000000840	!8<0000000000000840	!16<0000000000000840

= Strings =
string.pack("c5", "a") =	!>6100000000["a\0\0\0\0"]	!1>6100000000["a\0\0\0\0"]	!2>6100000000["a\0\0\0\0"]	!4>6100000000["a\0\0\0\0"]	!8>6100000000["a\0\0\0\0"]	!16>6100000000["a\0\0\0\0"]	!<6100000000["a\0\0\0\0"]	!1<6100000000["a\0\0\0\0"]	!2<6100000000["a\0\0\0\0"]	!4<6100000000["a\0\0\0\0"]	!8<6100000000["a\0\0\0\0"]	!16<6100000000["a\0\0\0\0"]
string.pack("c5", "abc") =	!>6162630000["abc\0\0"]	!1>6162630000["abc\0\0"]	!2>6162630000["abc\0\0"]	!4>6162630000["abc\0\0"]	!8>6162630000["abc\0\0"]	!16>6162630000["abc\0\0"]	!<6162630000["abc\0\0"]	!1<6162630000["abc\0\0"]	!2<6162630000["abc\0\0"]	!4<6162630000["abc\0\0"]	!8<6162630000["abc\0\0"]	!16<6162630000["abc\0\0"]
string.pack("c5", "abcde") =	!>6162636465	!1>6162636465	!2>6162636465	!4>6162636465	!8>6162636465	!16>6162636465	!<6162636465	!1<6162636465	!2<6162636465	!4<6162636465	!8<6162636465	!16<6162636465
string.pack("z", "a") =	!>6100	!1>6100	!2>6100	!4>6100	!8>6100	!16>6100	!<6100	!1<6100	!2<6100	!4<6100	!8<6100	!16<6100
string.pack("z", "abc") =	!>61626300	!1>61626300	!2>61626300	!4>61626300	!8>61626300	!16>61626300	!<61626300	!1<61626300	!2<61626300	!4<61626300	!8<61626300	!16<61626300
string.pack("z", "abcde") =	!>616263646500	!1>616263646500	!2>616263646500	!4>616263646500	!8>616263646500	!16>616263646500	!<616263646500	!1<616263646500	!2<616263646500	!4<616263646500	!8<616263646500	!16<616263646500
string.pack("s4", "a") =	!>0000000161	!1>0000000161	!2>0000000161	!4>0000000161	!8>0000000161	!16>0000000161	!<0100000061	!1<0100000061	!2<0100000061	!4<0100000061	!8<0100000061	!16<0100000061
string.pack("s4", "abc") =	!>00000003616263	!1>00000003616263	!2>00000003616263	!4>00000003616263	!8>00000003616263	!16>00000003616263	!<03000000616263	!1<03000000616263	!2<03000000616263	!4<03000000616263	!8<03000000616263	!16<03000000616263
string.pack("s4", "abcde") =	!>000000056162636465	!1>000000056162636465	!2>000000056162636465	!4>000000056162636465	!8>000000056162636465	!16>000000056162636465	!<050000006162636465	!1<050000006162636465	!2<050000006162636465	!4<050000006162636465	!8<050000006162636465	!16<050000006162636465
string.pack("s1", "a") =	!>0161	!1>0161	!2>0161	!4>0161	!8>0161	!16>0161	!<0161	!1<0161	!2<0161	!4<0161	!8<0161	!16<0161
string.pack("s1", "abc") =	!>03616263	!1>03616263	!2>03616263	!4>03616263	!8>03616263	!16>03616263	!<03616263	!1<03616263	!2<03616263	!4<03616263	!8<03616263	!16<03616263
string.pack("s1", "abcde") =	!>056162636465	!1>056162636465	!2>056162636465	!4>056162636465	!8>056162636465	!16>056162636465	!<056162636465	!1<056162636465	!2<056162636465	!4<056162636465	!8<056162636465	!16<056162636465

= Padding =
string.pack("BXjB", 255, 127) =	!>ff000000000000007f	!1>ff7f	!2>ff007f	!4>ff0000007f	!8>ff000000000000007f	!16>ff000000000000007f	!<ff000000000000007f	!1<ff7f	!2<ff007f	!4<ff0000007f	!8<ff000000000000007f	!16<ff000000000000007f
string.pack("BxB", 255, 127) =	!>ff007f	!1>ff007f	!2>ff007f	!4>ff007f	!8>ff007f	!16>ff007f	!<ff007f	!1<ff007f	!2<ff007f	!4<ff007f	!8<ff007f	!16<ff007f
